### PR TITLE
Update ftl-build container to Alpine v3.21

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Update `ftl-build` container to Alpine v3.21

HIGHLIGHTS:

- Linux kernel [6.12](https://kernelnewbies.org/Linux_6.12)
- GCC [14](https://gcc.gnu.org/gcc-14/changes.html)
- LLVM (clang) [19](https://releases.llvm.org/19.1.0/docs/ReleaseNotes.html)

As always, the major upgrade to GCC 14.2.0 is the most important reason for us to upgrade our build environment.

As it turned out that quite a few changes were necessary to support Alpine 3.12, I decided it's better to pin the container to a certain version instead of the more generic `latest`. It can always be updated and we can use it to tag a new container when new Alpine versions are released. The PR with the necessary changes for FTL will be opened once this PR is merged and new containers are tagged to update the used build container at the same time in the FTL PR.

---

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

File: ftl-build/Dockerfile [hunk 1]
```diff
@@ -45,9 +45,12 @@ RUN apk add --no-cache \
 ENV STATIC true
 ENV TEST true
 
+# As of Alpine 3.21 (Dec 2024), we need to patch the termcap library to include
+# the standard headers as well as unistd.h for the write function
 RUN curl -sSL https://ftl.pi-hole.net/libraries/termcap-${termcapversion}.tar.gz | tar -xz \
     && cd termcap-${termcapversion} \
     && ./configure --enable-static --disable-shared --disable-doc --without-examples \
+    && sed -i '1i #define STDC_HEADERS 1\n#include <unistd.h>' termcap.c tparam.c \
```
Comment (DL6ER) [inline ftl-build/Dockerfile:53]: Patching is necessary due to a [breaking change in GCC 14](https://gcc.gnu.org/gcc-14/porting_to.html#warnings-as-errors):

> **Implicit function declarations (`-Werror=implicit-function-declaration`)**
> It is no longer possible to call a function that has not been declared. In general, the solution is to include a header file with an appropriate function prototype. Note that GCC will perform further type checks based on the function prototype, which can reveal further type errors that require additional changes.
> 
> For well-known functions declared in standard headers, GCC provides fix-it hints with the appropriate `#include` directives:
> ```
>error: implicit declaration of function ‘strlen’ [-Wimplicit-function-declaration]
>    5 |   return strlen (s);
>      |          ^~~~~~
>note: include ‘<string.h>’ or provide a declaration of ‘strlen’
>  +++ |+#include <string.h>
>    1 |
> ```
> On GNU systems, headers described in standards (such as the C standard, or POSIX) may require the definition of certain macros at the start of the compilation before all required function declarations are made available. See [Feature Test Macros](https://sourceware.org/glibc/manual/latest/html_node/Feature-Test-Macros.html) in the GNU C Library manual for details.

**Comment by yubiuser** [comment]
Maybe we should also port the monkey patch from
https://github.com/pi-hole/FTL/pull/2044/commits
for nicer BATS output

**Comment by DL6ER** [comment]
@yubiuser As `TERM` is now set by us in the `tester`, we get this:
image

Anything needed in addition?

**Comment by yubiuser** [comment]
I thought about this part

https://github.com/pi-hole/FTL/blob/18f91b0f09c651841a3c82f03d7790c4fd482fd6/.github/Dockerfile#L19-L22

File: ftl-build/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-ARG CONTAINER="alpine:latest"
+ARG CONTAINER="alpine:3.21"
```
Comment (dschaper) [inline ftl-build/Dockerfile:1]: 👍🏻

**Review by dschaper — APPROVED** [review]

**Review by yubiuser — APPROVED** [review]

**Pull request merged**